torch/csrc/jit/passes/tensorexpr_fuser.cpp

```diff
@@ -190,118 +190,152 @@ Node* getOrCreateTensorExprSubgraph(Node* n) {
+    for (auto it = reverse_iter.begin(); it != reverse_iter.end();) {
+      Node* n = *it;
+
+      for (auto b : n->blocks()) {
+        createFusionGroups(b);
+      }
+
+      if (!canHandle(n)) {
+        it++;
+        continue;
+      }
+      // There are some nodes that we can support, but we don't want to start a
+      // fusion group from - skip them.
+      if (n->kind() == prim::ListConstruct || n->kind() == aten::slice ||
+          n->kind() == aten::unsqueeze || n->kind() == prim::ConstantChunk ||
+          n->kind() == prim::Constant) {
+        it++;
+        continue;
+      }
 
-    if (!canMerge(consumer, producer)) {
-      return c10::nullopt;
+      Node* fusion_group = createFusionGroup(n);
+      fusion_groups.push_back(fusion_group);
+      it = fusion_group->reverseIterator();
+      it++;
     }
 
-    consumer = getOrCreateTensorExprSubgraph(consumer);
+    for (auto n : fusion_groups) {
+      inlineIfTooSmall(n);
+    }
+  }
 
-    if (producer->kind() == aten::cat) {
-      Node* listconstruct = producer->inputs()[0]->node();
-
-      aliasDb_->moveBeforeTopologicallyValid(producer, consumer);
-      GRAPH_UPDATE(
-          "Merging ", getHeader(producer), " into ", getHeader(consumer));
-      SubgraphUtils::mergeNodeIntoSubgraph(producer, consumer);
-
-      aliasDb_->moveBeforeTopologicallyValid(listconstruct, consumer);
-      GRAPH_UPDATE(
-          "Merging ", getHeader(listconstruct), " into ", getHeader(consumer));
-      SubgraphUtils::mergeNodeIntoSubgraph(listconstruct, consumer);
-    } else {
-      aliasDb_->moveBeforeTopologicallyValid(producer, consumer);
-      GRAPH_UPDATE(
-          "Merging ", getHeader(producer), " into ", getHeader(consumer));
-      SubgraphUtils::mergeNodeIntoSubgraph(producer, consumer);
+  size_t blockSize(Block* block) {
+    size_t num = 0;
+    for (Node* n : block->nodes()) {
+      if (n->kind() == prim::Constant || n->kind() == prim::ListConstruct) {
+        continue;
+      }
+      for (Block* b : n->blocks()) {
+        num += blockSize(b);
+      }
+      num++;
+    }
+    return num;
+  }
+
+  bool inlineIfTooSmall(Node* n) {
+    AT_ASSERT(n->kind() == getTensorExprSymbol());
+    auto subgraph = SubgraphUtils::getSubgraph(n);
+    size_t num_modes = blockSize(subgraph->block());
+    if (num_modes < minSubgraphSize_) {
+      GRAPH_UPDATE("Fusion group is too small, unmerging: ", *n);
+      SubgraphUtils::unmergeSubgraph(n);
+      return true;
+    }
+    return false;
+  }
+
+  bool tryMerge(Node* fusion_group, Node* to_merge) {
+    if (!canMerge(fusion_group, to_merge)) {
+      return false;
     }
 
-    return consumer;
+    std::vector<Node*> nodes_to_merge = {to_merge};
+
+    if (to_merge->kind() == aten::cat) {
+      Node* listconstruct = to_merge->input(0)->node();
+      nodes_to_merge.push_back(listconstruct);
+    }
+    for (auto n : nodes_to_merge) {
+      aliasDb_->moveBeforeTopologicallyValid(n, fusion_group);
```

Comment:
That's a very good point and I'm trying to think how to properly check that without adding too much ugliness to the code.

Reply:
You can just check that `aliasDb_->moveBeforeTopologicallyValid(n, fusion_group);` is true, and bail if it's False

Reply:
It's a bit trickier than that because if it fails for ListConstruct we also want to not include the corresponding aten::cat. Currently the only way to properly do this imho is:
1) tryMove all chained nodes next to the fusion group, bail if any of the moves failed.
2) merge them into the group

We will have a side effect of moving things that we don't fuse, but I think it's a lesser evil in this case.

Reply:
I ended up splitting the loop where we move nodes to the fusion group and where we actually merge them into the group - I think it should handle the case you brought up correctly with the side effect mentioned in the previous comment.